### PR TITLE
Provide documentation for VerifyingKey and SigningKey

### LINKED ISSUE #56: Python 3: SigningKey.to_string() return a bytes object instead of a string
```
>>> import ecdsa
>>> type(ecdsa.SigningKey.generate().to_string())
<class 'bytes'>
```

Is this normal?

Example of how to get a string for someone that need to display a key:

```
>>> import codecs
>>> sk = ecdsa.SigningKey.generate().to_string()
>>> codecs.encode(sk, 'hex').decode("utf-8")  # maybe there is a simpler way
'5d9411d69b0286f51d5ef52300dcd02b617d843877c8a881'
```

### PR DESCRIPTION
**Pull request by tomato42**
Document the VerifyingKey, SigningKey and few directly related classes.

fixes #56 

work towards #95

### CODE REVIEW

**Comment by tomato42** [comment]
@ansasaki please review

File: src/ecdsa/keys.py [hunk 1]
```diff
@@ -222,21 +472,112 @@ def to_string(self, encoding="raw"):
...
+
+        The PEM header of the key will be `BEGIN PUBLIC KEY`.
+
+        The format of the key is described in the `from_der()` method.
+        This method supports only "named curve" encoding of keys.
+
+        :param str point_encoding: specification of the encoding format
+            of public keys. "uncompressed" is most portable, "compressed" is
+            smallest, "hybrid" is uncommon and unsupported by most
+            implementations, is as big as "uncompressed". "raw" will create
```
Comment (ansasaki) [inline src/ecdsa/keys.py]: Why is "raw" allowed if it generates malformed files?

File: src/ecdsa/keys.py [hunk 2]
```diff
@@ -222,21 +472,112 @@ def to_string(self, encoding="raw"):
...
 
-    def verify(self, signature, data, hashfunc=None, sigdecode=sigdecode_string):
+    def verify(self, signature, data, hashfunc=None,
+               sigdecode=sigdecode_string):
+        """
+        Verify a signature made over provided data.
+
+        Will hash `data` to verify the signature.
+
+        By default expects signature in raw encoding. Can also be used to
```
Comment (ansasaki) [inline src/ecdsa/keys.py]: Should ':term:`raw encoding`' be used here?

File: src/ecdsa/keys.py [hunk 3]
```diff
@@ -222,21 +472,112 @@ def to_string(self, encoding="raw"):
...
+        """
         hashfunc = hashfunc or self.default_hashfunc
         digest = hashfunc(data).digest()
         return self.verify_digest(signature, digest, sigdecode)
 
     def verify_digest(self, signature, digest, sigdecode=sigdecode_string):
+        """
+        Verify a signature made over provided hash value.
+
+        By default expects signature in raw encoding. Can also be used to
```
Comment (ansasaki) [inline src/ecdsa/keys.py]: Should ':term:raw encoding' be used here?

File: src/ecdsa/util.py [hunk 4]
```diff
@@ -212,13 +212,43 @@ def sigencode_strings(r, s, order):
...
 
 def sigencode_string(r, s, order):
+    """
+    Encode the signature to raw format.
+
+    :param int r: first parametr of the signature
+    :param int s: second parameter of the signature
+    :param int order: the order of curve over which the signature was computed
+
+    :return: DER encoding of ECDSA signature
```
Comment (ansasaki) [inline src/ecdsa/util.py]: It returns the raw encoding or DER encoding?

File: src/ecdsa/util.py [hunk 5]
```diff
@@ -244,44 +274,113 @@ def sigencode_der_canonize(r, s, order):
...
     r = string_to_number_fixedlen(signature[:l], order)
     s = string_to_number_fixedlen(signature[l:], order)
     return r, s
 
 
 def sigdecode_strings(rs_strings, order):
+    """
+    Decode the signature from two strings.
+
+    First string needs to encode 'r', second needs to encode the
```
Comment (ansasaki) [inline src/ecdsa/util.py]: Encode how? In other parts the expected encoding was more specific.

File: src/ecdsa/keys.py [hunk 6]
```diff
@@ -10,35 +70,80 @@
...
+        self.pubkey = None
 
     @classmethod
-    def from_public_point(klass, point, curve=NIST192p, hashfunc=sha1):
-        self = klass(_error__please_use_generate=True)
+    def from_public_point(cls, point, curve=NIST192p, hashfunc=sha1):
+        """
+        Initialise the object from a Point object.
+
+        This is a low-level method, generally you will not want to use it.
```
[the remark this answers is not in the document]
Comment (tomato42) [reply src/ecdsa/keys.py]: both: programmer can decode the point himself or herself, but it's error-prone; generally the point will be encoded somehow, so the from_der, from_pem, from_string are more appropriate...

File: src/ecdsa/keys.py (hunk 1 above)
Comment (tomato42) [reply src/ecdsa/keys.py]: because there is no code to forbid it, filed #136 to fix that (I want to keep this PR docs and stylistic changes only)

**Comment by tomato42** [comment]
thanks for review; will fix the issues I haven't commented on

**Comment by tomato42** [comment]
@ansasaki updated, please re-check

File: src/ecdsa/util.py [hunk 7]
```diff
@@ -212,13 +212,43 @@ def sigencode_strings(r, s, order):
 
 
 def sigencode_string(r, s, order):
+    """
+    Encode the signature to raw format (:term:`raw encoding`)
+
+    :param int r: first parametr of the signature
```
Comment (ansasaki) [inline src/ecdsa/util.py]: s/parametr/parameter/

File: src/ecdsa/util.py [hunk 8]
```diff
@@ -212,13 +212,43 @@ def sigencode_strings(r, s, order):
...
+    :param int s: second parameter of the signature
+    :param int order: the order of curve over which the signature was computed
```
Comment (ansasaki) [inline src/ecdsa/util.py]: s/the order of curve/the order of the curve/

File: src/ecdsa/util.py [hunk 9]
```diff
@@ -212,13 +212,43 @@ def sigencode_strings(r, s, order):
...
+
+        Ecdsa-Sig-Value ::= SEQUENCE {
+            r       INTEGER,
+            s       INTEGER
+        }
+
+    It's expected that this function will be used as a `sigencode=` parameter
+    in SigningKey.sign() method.
+
+    :param int r: first parametr of the signature
```
Comment (ansasaki) [inline src/ecdsa/util.py]: s/parametr/parameter

File: src/ecdsa/util.py [hunk 10]
```diff
@@ -212,13 +212,43 @@ def sigencode_strings(r, s, order):
...
+    :param int s: second parameter of the signature
+    :param int order: the order of curve over which the signature was computed
```
Comment (ansasaki) [inline src/ecdsa/util.py]: s/the order of curve/the order of the curve/

File: src/ecdsa/keys.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ansasaki) [reply src/ecdsa/keys.py]: Thank you for the clarification.

**Comment by tomato42** [comment]
> I found some typos.

@ansasaki good catch, can you verify?

**Comment by tomato42** [comment]
(I've also "sphinxified" the cross-links and made them a bit more consistent)

File: src/ecdsa/keys.py [hunk 11]
```diff
@@ -1,5 +1,65 @@
...
+
+    DER
+        Distinguished Encoding Rules, the way to encode :term:`ASN.1` objects
+        deterministically and uniquely into byte strings.
+
+    ASN.1
+        Abstract Syntax Notation 1 is a standard description language for
+        specifying serialisation and deserialisation of data structures in a
+        portable and cross-platform way.
+"""
```
Comment (ansasaki) [inline src/ecdsa/keys.py]: Just a nit, the glossary could be alphabetically sorted.

File: src/ecdsa/keys.py [hunk 12]
```diff
@@ -153,29 +337,79 @@ def from_der(klass, string):
         if empty != b(""):
             raise der.UnexpectedDER("trailing junk after pubkey pointstring: %s" %
                                     binascii.hexlify(empty))
-        # the point encoding is padded with a zero byte
-        # raw encoding of point is invalid in DER files
-        if not point_str.startswith(b("\x00")) or \
-                len(point_str[1:]) == curve.verifying_key_length:
+        # the bitsting encoding is padded with a zero byte
```
Comment (ansasaki) [inline src/ecdsa/keys.py]: s/bitsting/bitstring/

File: src/ecdsa/keys.py (hunk 11 above)
Comment (tomato42) [reply src/ecdsa/keys.py]: it will be once it is processed by sphinx :)

**Comment by tomato42** [comment]
Thank you!

**Pull request merged**